File: js/src/core/attestation-model.ts
import type {
  GpuEvidenceStatus,
  DeploymentVerifier,
  GpuEvidenceVerifier,
  VerifiedModelAttestation,
  VerifyModelAttestationParams,
} from '../types/verification';
import { Buffer } from 'buffer';
import { computeAddress } from 'ethers';
import { decodeNvidiaPayloadNonce } from '../boundaries/nvidia';
import type { SigningAlgo } from '../types/attestation-common';
import type { ModelAttestation } from '../types/attestation-model';
import { hexToBuffer } from '../utils/common';
import { VerificationError, wrapVerificationError } from '../utils/errors';
import { createGpuEvidenceVerifier } from '../utils/nvidia';
import {
  verifyReportDataBinding,
  verifyReportedNonce,
} from './attestation-common';
import {
  verifyDstackDeployment,
  verifyDstackQuote,
} from './dstack-attestation';
import type { VerifiedDstackQuote } from './dstack-attestation';

/**
 * Verify model evidence returned through NEAR AI Cloud. This verifies freshness
 * and the model signing identity but does not claim a client-to-model TLS
 * binding; the client's TLS connection terminates at the gateway.
 */
export async function verifyModelAttestation(
  params: VerifyModelAttestationParams,
): Promise<VerifiedModelAttestation> {
  // CPU and GPU evidence bind independently to the same client nonce.
  const [deployment, gpuEvidence] = await Promise.all([
    verifyModelCpuAttestation(params),
    verifyModelGpuEvidence(params),
  ]);
  return { ...deployment, gpuEvidence };
}

type VerifiedModelDeployment = Omit<VerifiedModelAttestation, 'gpuEvidence'>;

async function verifyModelCpuAttestation({
  attestation,
  clientBinding,
  policy,
  verifiers,
}: VerifyModelAttestationParams): Promise<VerifiedModelDeployment> {
  const { nonce } = clientBinding;
  const verifiedQuote = await verifyDstackQuote({
    attestation,
    nonce,
    policy,
    tdxQuoteVerifier: verifiers?.tdxQuote,
    advertisedReportData: attestation.reportedQuoteData,
  });
  verifyReportDataBinding({
    reportData: verifiedQuote.quote.reportData,
    nonce,
    signingAddress: verifiedQuote.signer.signingAddress,
  });
  return verifyModelDeployment({
    attestation,
    verifiedQuote,
    deploymentVerifier: verifiers?.deployment,
  });
}

type VerifyModelDeploymentParams = {
  attestation: ModelAttestation;
  verifiedQuote: VerifiedDstackQuote;
  deploymentVerifier?: DeploymentVerifier;
};

/** Shared model checks after the endpoint-specific report-data binding passes. */
export async function verifyModelDeployment({
  attestation,
  verifiedQuote,
  deploymentVerifier,
}: VerifyModelDeploymentParams): Promise<VerifiedModelDeployment> {
  const evidence = await verifyDstackDeployment(
    verifiedQuote,
    deploymentVerifier,
  );

  const signingPublicKey = verifySigningPublicKey({
    attestation,
    signingAlgo: evidence.signer.signingAlgo,
    signingAddress: evidence.signer.signingAddress,
  });

  return {
    ...evidence,
    ...(signingPublicKey === undefined ? {} : { signingPublicKey }),
  };
}

/** GPU verification is independent of the model's CPU quote and deployment. */
export function verifyModelGpuEvidence({
  attestation,
  clientBinding: { nonce },
  policy,
  verifiers,
}: VerifyModelAttestationParams): Promise<GpuEvidenceStatus> {
  return verifyGpuEvidence({
    payload: attestation.nvidiaPayload,
    nonce,
    requirement: policy?.gpuEvidence ?? 'if-present',
    verifier: verifiers?.gpuEvidence ?? createGpuEvidenceVerifier(),
  });
}

type VerifySigningPublicKeyParams = {
  readonly attestation: VerifyModelAttestationParams['attestation'];
  readonly signingAlgo: SigningAlgo;
  readonly signingAddress: string;
};

/**
 * When Cloud API supplies a model E2EE public key, bind it to the signer
 * already authenticated by quote report data. Ed25519 signer addresses are
 * the public key itself. ECDSA signer addresses are derived from the raw
 * secp256k1 `X || Y` public key.
 */
function verifySigningPublicKey({
  attestation,
  signingAlgo,
  signingAddress,
}: VerifySigningPublicKeyParams): string | undefined {
  if (attestation.signingPublicKey === undefined) {
    return undefined;
  }
  const signingPublicKey = hexToBuffer(
    attestation.signingPublicKey,
    'attestation.signingPublicKey',
  );
  const verifiedSigningAddress = hexToBuffer(
    signingAddress,
    'signer.signingAddress',
  );
  const publicKeyMatchesSigner =
    signingAlgo === 'ed25519'
      ? signingPublicKey.length === 32 &&
        signingPublicKey.equals(verifiedSigningAddress)
      : ecdsaPublicKeyMatchesSigner(signingPublicKey, verifiedSigningAddress);
  if (!publicKeyMatchesSigner) {
    throw new VerificationError({
      code: 'binding.model_public_key_mismatch',
    });
  }
  return signingAlgo === 'ecdsa' && signingPublicKey.length === 65
    ? signingPublicKey.subarray(1).toString('hex')
    : signingPublicKey.toString('hex');
}

function ecdsaPublicKeyMatchesSigner(
  signingPublicKey: Buffer,
  signingAddress: Buffer,
): boolean {
  const rawPublicKey =
    signingPublicKey.length === 65 && signingPublicKey[0] === 0x04
      ? signingPublicKey.subarray(1)
      : signingPublicKey;
  if (rawPublicKey.length !== 64) {
    return false;
  }

  try {
    const derivedSigningAddress = Buffer.from(
      computeAddress(`0x04${rawPublicKey.toString('hex')}`).slice(2),
      'hex',
    );
    return derivedSigningAddress.equals(signingAddress);
  } catch {
    return false;
  }
}

type VerifyGpuEvidenceParams = {
  payload?: string;
  nonce: string;
  requirement: 'if-present' | 'required';
  verifier: GpuEvidenceVerifier;
};

async function verifyGpuEvidence(
  input: VerifyGpuEvidenceParams,
): Promise<GpuEvidenceStatus> {
  if (input.payload === undefined) {
    if (input.requirement === 'required') {
      throw new VerificationError({
        code: 'policy.gpu_evidence_required',
      });
    }
    return 'not_provided';
  }

  // Bind the provider payload to the same nonce before handing it to either
  // the default NRAS verifier or a caller-supplied GPU verifier.
  verifyReportedNonce({
    reportedNonce: decodeNvidiaPayloadNonce(input.payload),
    nonce: input.nonce,
    source: 'nvidiaPayload',
  });

  try {
    await input.verifier(input.payload);
  } catch (cause) {
    throw wrapVerificationError(
      {
        code: 'gpu.attestation_rejected',
        details: { source: 'custom_verifier' },
      },
      cause,
    );
  }
  return 'verified';
}

File: py/src/nearai_inference_sdk/core/attestation_model.py
"""Model-serving TEE attestation verification."""

from __future__ import annotations

from cryptography.hazmat.primitives.asymmetric import ec
from eth_utils import keccak

from ..types.verification import (
    GpuEvidenceVerifier,
    ModelAttestationPolicy,
    ModelAttestationVerifiers,
    ModelClientBinding,
    VerifiedModelAttestation,
)
from ..types.attestation_model import ModelAttestation
from ..utils.common import hex_to_bytes, maybe_await
from ..utils.errors import (
    VerificationError,
    verification_failure,
)
from ..utils.nvidia import decode_nvidia_payload_nonce, verify_nvidia_nras
from .attestation_common import verify_report_data_binding, verify_reported_nonce
from .dstack_attestation import verify_dstack_deployment, verify_dstack_quote


async def verify_model_attestation(
    attestation: ModelAttestation,
    client_binding: ModelClientBinding,
    *,
    policy: ModelAttestationPolicy | None = None,
    verifiers: ModelAttestationVerifiers | None = None,
) -> VerifiedModelAttestation:
    """Verify model evidence returned through NEAR AI Cloud."""

    nonce = client_binding.nonce
    verified_quote = await verify_dstack_quote(
        attestation=attestation,
        advertised_report_data=attestation.reported_quote_data,
        nonce=nonce,
        policy=policy,
        tdx_quote_verifier=None if verifiers is None else verifiers.tdx_quote,
    )
    verify_report_data_binding(
        report_data=verified_quote.quote.report_data,
        nonce=nonce,
        signer=verified_quote.signer,
    )
    evidence = await verify_dstack_deployment(
        verified_quote, None if verifiers is None else verifiers.deployment
    )
    gpu_evidence = await _verify_gpu_evidence(
        payload=attestation.nvidia_payload,
        nonce=nonce,
        policy=policy,
        verifier=None if verifiers is None else verifiers.gpu_evidence,
    )
    return VerifiedModelAttestation(
        signer=evidence.signer,
        tcb_status=evidence.tcb_status,
        advisory_ids=evidence.advisory_ids,
        deployment=evidence.deployment,
        deployment_provenance=evidence.deployment_provenance,
        gpu_evidence=gpu_evidence,
        signing_public_key=_verify_signing_public_key(attestation),
    )


def _verify_signing_public_key(attestation: ModelAttestation) -> str | None:
    """Bind an optional E2EE key to the quote-authenticated model signer."""

    if attestation.signing_public_key is None:
        return None
    public_key = hex_to_bytes(
        attestation.signing_public_key, 'attestation.signing_public_key'
    )
    address = hex_to_bytes(attestation.signer.signing_address, 'signer.signing_address')
    if attestation.signer.signing_algo == 'ed25519':
        matches = len(public_key) == 32 and public_key == address
    else:
        if len(public_key) == 65 and public_key[0] == 4:
            public_key = public_key[1:]
        matches = False
        if len(public_key) == 64:
            try:
                ec.EllipticCurvePublicKey.from_encoded_point(
                    ec.SECP256K1(), b'\x04' + public_key
                )
                matches = keccak(public_key)[-20:] == address
            except ValueError:
                pass
    if not matches:
        raise verification_failure('binding.model_public_key_mismatch')
    return public_key.hex()


async def _verify_gpu_evidence(
    *,
    payload: str | None,
    nonce: str,
    policy: ModelAttestationPolicy | None,
    verifier: GpuEvidenceVerifier | None,
) -> str:
    requirement = 'if-present' if policy is None else policy.gpu_evidence
    if payload is None:
        if requirement == 'required':
            raise verification_failure('policy.gpu_evidence_required')
        return 'not_provided'

    payload_nonce = decode_nvidia_payload_nonce(payload)
    verify_reported_nonce(payload_nonce, nonce, 'nvidiaPayload')

    try:
        if verifier is None:
            await verify_nvidia_nras(payload, nonce)
        else:
            await maybe_await(verifier(payload))
    except VerificationError:
        raise
    except Exception as error:
        raise verification_failure(
            'gpu.attestation_rejected',
            {'source': 'custom_verifier'},
            cause=error,
        ) from error
    return 'verified'
